src/rulesets/Base/arraymath.jl

```diff
@@ -19,28 +19,49 @@ end
 ##### `*`
 #####
 
+
 function rrule(
     ::typeof(*),
     A::AbstractVecOrMat{<:CommutativeMulNumber},
     B::AbstractVecOrMat{<:CommutativeMulNumber},
 )
     project_A = ProjectTo(A)
     project_B = ProjectTo(B)
+    function times_pullback(ȳ)
+        Ȳ = unthunk(ȳ)
+        dA = @thunk(project_A(Ȳ * B'))
+        dB = @thunk(project_B(A' * Ȳ))
+        return NoTangent(), dA, dB
+    end
+    return A * B, times_pullback
+end
+
+# Optimized case for StridedMatrixes
+# no need to project as already dense, and we are allowed to use InplaceableThunk because
+# we know the destination will also be dense. TODO workout how to apply this generally:
```

Comment:
No need to project for structure, but possibly for eltype, right?

Reply:
this is true. we might run into a problem with  mistakenly complixfying a real matrix.